[WIP] Add OpenCL runtime

Menooker opened the pull request:
This is a PR still work-in-progress. I have added opencl-runtime in some of the runner tests. It seems to work well. Todo:

- [ ] Polish error-handling
- [ ] OpenCL distinguish pointer args with non-pointer args. I need to check if an arg is pointer every time the kernel is called. Is there any way to bypass it?
- [ ] add opencl-runtime in most of runner tests.

silee2 commented:
I know that the PR is still WIP but left some small comments. Mostly about name usage.

Menooker commented:
> I know that the PR is still WIP but left some small comments. Mostly about name usage.

Thanks for the comments. I have updated the code now. Those names were copy-pasted from L0 wrapper and I have now changed them accordingly. I have added opencl-runtime for most of the UTs. And I found some of the tests failing. Still working on it...

Menooker commented on lib/ExecutionEngine/OPENCLRUNTIME/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set(CMAKE_CXX_FLAGS " ${CMAKE_CXX_FLAGS} -Wno-return-type -Wno-cast-qual")
```
This flag ends up unnecessary. I have removed this flag.

Menooker commented:
Removed the "WIP" status and this PR is ready to review. Added additional RUN in UTs for OpenCL runtime.

> And I found some of the tests failing. Still working on it...

The error ends up the wrong setting of WG size. L0, Sycl and Ocl have different ways to set the WG size and x/y/x orders. :(
